rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+workflow:
+
+.. code::python
+
+    bb = client.create_batch()
+    job_tree = bb.job_tree()
+    session = job_tree.create_path('/session1')
+    driver = session.create_job(name='driver')
+    stage1 = job_tree.create_path('/session1/stage1')
+    for i in range(5):
+        stage1.create_job(name=f'worker{i}')
+    b = bb.submit()
+
+Oh no! The query is taking too long. Let's cancel the entire session,
+but not the batch in case there's multiple simultaneous queries
+happening:
+
+.. code::python
+
+   session = b.job_tree().get_path('/session1')
+   session.cancel()
+   session.wait()
+
+A user wants to track how much it costs to run the PCA part of the
+pipeline for multiple queries:
+
+.. code::python
+
+    bb = client.create_batch()
+    job_tree = bb.job_tree()
+    session = job_tree.create_path('/session1')
+    driver = session.create_job(name='driver')
+
+    stage1 = job_tree.create_path('/session1/stage1')
+    for i in range(5):
+        stage1.create_job(name=f'worker{I}', attributes={'pca': '1'})
+
+    stage2 = job_tree.create_path('/session1/stage2')
+    for i in range(10):
+        stage2.create_job(name=f'worker{I}', attributes={'pca': '1'})
+
+    stage3 = job_tree.create_path('/session1/stage3')
+    for i in range(10):
+        stage3.create_job(name=f'worker{I}', attributes={'vep': '1'})
+
+    b = bb.submit()
+    b.wait()
+
+    pca = b.create_job_group('"pca"')
+    status = pca.status()
+    pca_cost = status['cost']
+
+
+Finally, let's select the jobs in that group that cost more than $5
+each:
+
+.. code::python
+
+    for j in pca.list_jobs('cost > 5'):
+        print(j)
+
+
+For completeness, if we want to manually add jobs to an arbitrary
+preexisting job group, we can do the following. However, I don't think
+this will be a common use case and we can implement it **later** on:
+
+.. code::python
+
+    bb = client.create_batch()
+    for i in range(5):
+        bb.create_job(name=f'worker{i}')
+    b = bb.submit()
+
+    random_jg = client.create_job_group(b.id, 'random')
+    for j in b.list_jobs():
+        if random.random() > 0.5:
+            random_jg.add_job(j['job_id'])
+    random_jg.update()
+
+
+-----------------------
+Effect and Interactions
+-----------------------
+
+My proposed changes address the issues raised in the motivation by
+providing the following features:
+
+1. Expose a way to quickly cancel a subset of jobs in a batch.
+2. Expose a way to quickly cancel a subset of jobs in a batch after a
+   specified number of failures in the group.
+3. Expose a way to quickly find the cost and status of a subset of
+   jobs in a batch.
+4. Expose a tree hierarchy structure for jobs to improve the user
+   experience in both the UI and for QoB interactive sessions.
+
+There are no interactions with existing features. This feature
+proposal is purely an addition to what we have in our system currently
+and maintains backwards compatibility.
```

Comment:
This feels weird to say. It does affect things like marking jobs complete. It adds work. I'm not sure this paragraph is necessary.